<?xml version="1.0"?>
<!-- Revolute-Revolute Manipulator -->
<robot name="robotino_base" xmlns:xacro="http://www.ros.org/wiki/xacro">

    <xacro:include filename="omni_wheels.urdf.xacro" />
    <xacro:include filename="robotino_base.gazebo.xacro" />

    <!-- Constants for robot dimensions -->
    <xacro:property name="PI" value="3.1415926535897931" />
    <xacro:property name="mass" value="1" /> <!-- arbitrary value for mass -->
    <xacro:property name="friction" value="0.7" /> <!-- Square dimensions (widthxwidth) of beams -->
    <xacro:property name="radius" value="0.176" /> <!-- Link 1 -->
    <xacro:property name="height" value="0.150" /> <!-- Link 2 -->
    <xacro:property name="wheel_radius" value="0.0375" /> <!-- Link 1 -->
    <xacro:property name="wheel_height" value="0.06" /> <!-- Link 2 -->

    <!-- Base Link -->
    <link name="base_link">
    </link>

    <link name="structure_link">
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <cylinder radius="${radius}" length="${height}" />
            </geometry>
        </collision>

        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <cylinder radius="${radius}" length="${height}" />
            </geometry>
            <material name="white">
                <color rgba="1 0.5 1 1" />
            </material>
        </visual>

        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <mass value="${mass}" />
            <inertia
                ixx="0.5" ixy="0.0" ixz="0.0"
                iyy="0.5" iyz="0.0"
                izz="0.5" />
        </inertial>
    </link>

    <link name="lidar_link">
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.05" length="0.02" />
            </geometry>
        </collision>

        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.05" length="0.02" />
            </geometry>
            <material name="black">
                <color rgba="0 0 0 1" />
            </material>
        </visual>

        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <mass value="0.200" />
            <inertia
                ixx="0.1" ixy="0.0" ixz="0.0"
                iyy="0.1" iyz="0.0"
                izz="0.1" />
        </inertial>
    </link>

    <joint name="base_to_structure" type="fixed">
        <parent link="base_link" />
        <child link="structure_link" />
    </joint>

    <joint name="structure_to_lidar" type="fixed">
        <parent link="structure_link" />
        <child link="lidar_link" />
        <origin rpy="0 0 0" xyz="-0.1 -0.08 ${(height/2)+0.02}" />
    </joint>

    <xacro:omni_wheel suffix="back" radius="${wheel_radius}" height="${wheel_height}" mass="0.2">
        <origin xyz="-0.13 0 ${(-height/2)+0.01}" rpy="${PI/2} 0 ${PI/2}" />
        <parent link="structure_link" />
    </xacro:omni_wheel>

    <xacro:omni_wheel suffix="right" radius="${wheel_radius}" height="${wheel_height}" mass="0.2">
        <origin xyz="0.065 -0.11 ${(-height/2)+0.01}" rpy="${PI/2} 0 ${PI+PI/3}" />
        <parent link="structure_link" />
    </xacro:omni_wheel>

    <xacro:omni_wheel suffix="left" radius="${wheel_radius}" height="${wheel_height}" mass="0.2">
        <origin xyz="0.065 0.11 ${(-height/2)+0.01}" rpy="${PI/2} 0 ${-PI/3}" />
        <parent link="structure_link" />
    </xacro:omni_wheel>

    <ros2_control name="IgnitionSystem" type="system">
        <hardware>
            <plugin>ign_ros2_control/IgnitionSystem</plugin>
        </hardware>
        <joint name="omni_back_joint">
            <command_interface name="velocity">
                <param name="min">-15</param>
                <param name="max">15</param>
            </command_interface>
            <state_interface name="position" />
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>
        <joint name="omni_left_joint">
            <command_interface name="velocity">
                <param name="min">-15</param>
                <param name="max">15</param>
            </command_interface>
            <state_interface name="position" />
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>
        <joint name="omni_right_joint">
            <command_interface name="velocity">
                <param name="min">-15</param>
                <param name="max">15</param>
            </command_interface>
            <state_interface name="position" />
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>
    </ros2_control>

</robot>